### PR TITLE
Bach: Windowing rework

### PR DESCRIPTION
**Pull request by jansentom**
- use `df.groupby(cols).window()` instead `df.window(cols)`, mainly to make `rolling()` and `expanding()` more pandas-like.
    - Make window function `window` argument Optional
    - Window functions now unset `group_by` on the Series, because they can be evaluated without it.
    - in `agg()`, update the index of the new df by looking at the returned Series.
    - Update some docs
    - Use new windowing syntax in `product_analytics.ipynb` and `series_objectiv.py`

This allows for syntax like
```
# preferred way 
df['col'] = df.groupby('colA').rolling(2).sum()
# or 
rolling = df.groupby('colA').rolling(2)
df['col'] = df.colA.sum(rolling)

# and deprecates
df['col'] = df.rolling('colA', 2).sum()
```

`window()` has been marked as expert in the docs, and we point users to `rolling()` and `expanding()` which should do most tricks.

### CODE REVIEW

File: bach/tests/functional/bach/test_df_window.py [hunk 1]
```diff
@@ -270,43 +308,119 @@ def test_windowing_expressions():
...
+        r = df[['skating_order', 'municipality']]
+        # the first sort_order is for the window order
+        t = df.sort_values('skating_order')
+        t = t.groupby(group_by) if group_by else t
+
+        rolling = t.rolling(
+            window=window, min_periods=min_periods, center=center
+        ).inhabitants.sum()
+
+        if isinstance(df, DataFrame):
```
Comment (jansentom) [inline bach/tests/functional/bach/test_df_window.py:365]: Figuring this out was hard ;(

File: analysis/notebooks/product_analytics.ipynb [hunk 2]
```diff
@@ -636,9 +636,9 @@
     "# use the index to give each feature a unique id\n",
     "feature_id_pd['feature_id'] = feature_id_pd.index\n",
     "\n",
-    "# create a window that returns the previous event for each row\n",
-    "window = converting_users_events.sort_values('session_hit_number').window('session_id')\n",
-    "converting_users_events['prev_event_location'] = converting_users_events.event_location.window_lag(window)\n",
+    "# create a rolling window that includes the previous event for each row and get it through window_lag()\n",
```
Comment (jansentom) [inline analysis/notebooks/product_analytics.ipynb:639]: @vincenthoogsteder FYI more pandas-like syntax coming up here.

Comment (vincenthoogsteder) [reply analysis/notebooks/product_analytics.ipynb:639]: fancy

File: bach/tests/functional/bach/test_df_window.py [hunk 3]
```diff
@@ -270,43 +308,119 @@ def test_windowing_expressions():
...
+                for bdf_values, pdf_values in [
+                    (get_rolling_values(bdf) , get_rolling_values(pdf)),
+                    (
+                            get_rolling_values(bdf, ['municipality']),
+                            get_rolling_values(pdf, ['municipality']),
+                    )]:
+                    try:
+                        # Because the returned values have nan / None mixture,
+                        # it's really hard to compare them easily.
+                        for x, y in zip(bdf_values.tolist(), pdf_values.tolist()):
```
Comment (jansentom) [inline bach/tests/functional/bach/test_df_window.py:388]: actually, mostly this part was annoying. There is no nice way to deal with array of values, nan, None in any combination.

**Review by thijs-obj — APPROVED** [review]
Some small nitpicks, but nothing substantial

File: bach/bach/series/series.py [hunk 4]
```diff
@@ -815,13 +815,24 @@ def _check_unwrap_groupby(self,
         Exceptions will be raised when check don't pass
         :returns: The potentially unwrapped GroupBy
         """
-        from bach.partitioning import GroupBy
-        isin = GroupBy if isin is None else isin
+        from bach.partitioning import GroupBy, Cube, Rollup, Window
+        isin = (GroupBy, Cube, Rollup, Window) if isin is None else isin
```
[the remark this answers is not in the document]
Comment (jansentom) [reply bach/bach/series/series.py:819]: I think I refactored this a bit and then undid that. I fixed the comment to make sure it actually matches what the method does and undid the change for `GroupBy` subclass checks.

File: bach/tests/functional/bach/test_df_window.py [hunk 5]
```diff
@@ -270,43 +308,119 @@ def test_windowing_expressions():
...
+                            xnan = x is None or np.isnan(x)
+                            ynan = y is None or np.isnan(y)
+                            if xnan != ynan or (not xnan and x != y):
+                                # this look ridiculous. but actually gives a nice error
```
[the remark this answers is not in the document]
Comment (jansentom) [reply bach/tests/functional/bach/test_df_window.py:389]: YW! I fixed the typo too :)

File: bach/tests/functional/bach/test_df_window.py (hunk 3 above)
Comment (jansentom) [reply bach/tests/functional/bach/test_df_window.py:385]: Yeah, or just abstract this out into utils. One of those when I have to fix this next time

**Pull request merged**